// @authllizer/ngx/es5/authllizer.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var http_1 = require("@angular/common/http");
var core_2 = require("@authllizer/core");
var ngx_http_1 = require("./ngx.http");
var AuthllizerModule = /** @class */ (function () {
    function AuthllizerModule() {
    }
    AuthllizerModule.forRoot = function (options) {
        return {
            ngModule: AuthllizerModule,
            providers: [{
                    provide: core_2.Authllizer,
                    useFactory: function (http) {
                        options.httpClient = options.httpClient || new ngx_http_1.NgxHttpClient(http);
                        return new core_2.Authllizer(options);
                    },
                    deps: [http_1.HttpClient]
                }]
        };
    };
    AuthllizerModule.decorators = [
        { type: core_1.NgModule, args: [{
                    imports: [
                        http_1.HttpClientModule
                    ]
                },] },
    ];
    /** @nocollapse */
    AuthllizerModule.ctorParameters = function () { return []; };
    return AuthllizerModule;
}());
exports.AuthllizerModule = AuthllizerModule;
