// @authllizer/ngx/es5/ngx.http.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@authllizer/core");
var NgxHttpClient = /** @class */ (function (_super) {
    __extends(NgxHttpClient, _super);
    function NgxHttpClient(httpClient) {
        return _super.call(this, httpClient) || this;
    }
    NgxHttpClient.prototype.request = function (url, options) {
        var method = options.method, data = options.data, params = options.params, headers = options.headers, withCredentials = options.withCredentials;
        return this._client.request(method, url, {
            body: data,
            params: params,
            headers: headers,
            withCredentials: withCredentials
        }).toPromise();
    };
    return NgxHttpClient;
}(core_1.BaseHttpClient));
exports.NgxHttpClient = NgxHttpClient;
